peer/msg_router.go

```diff
@@ -0,0 +1,270 @@
+package peer
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/lightningnetwork/lnd/fn"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"golang.org/x/exp/maps"
+)
+
+var (
+	// ErrDuplicateEndpoint is returned when an endpoint is registered with
+	// a name that already exists.
+	ErrDuplicateEndpoint = fmt.Errorf("endpoint already registered")
+
+	// ErrUnableToRouteMsg is returned when a message is unable to be
+	// routed to any endpoints.
+	ErrUnableToRouteMsg = fmt.Errorf("unable to route message")
+)
+
+// EndPointName is the name of a given endpoint. This MUST be unique across all
+// registered endpoints.
+type EndPointName = string
+
+// MsgEndpoint is an interface that represents a message endpoint, or the
+// sub-system that will handle processing an incoming wire message.
+type MsgEndpoint interface {
+	// Name returns the name of this endpoint. This MUST be unique across
+	// all registered endpoints.
+	Name() EndPointName
+
+	// CanRouteTo returns true if the target message can be routed to this
+	// endpoint.
+	CanRouteTo(msg lnwire.Message) bool
+
+	// HandleMessage handles the target message, and returns true if the
+	// message was able to be processed.
+	SendMessage(msg lnwire.Message) bool
+}
+
+// MsgRouter is an interface that represents a message router, which is generic
+// sub-system capable of routing any incoming wire message to a set of
+// registered endpoints.
+//
+// TODO(roasbeef): move to diff sub-system?
+type MsgRouter interface {
+	// RegisterEndpoint registers a new endpoint with the router. If a
+	// duplicate endpoint exists, an error is returned.
+	RegisterEndpoint(MsgEndpoint) error
+
+	// UnregisterEndpoint unregisters the target endpoint from the router.
+	UnregisterEndpoint(EndPointName) error
```

Comment:
If we don't support unregistering, then how would we ever clean up the resources allocated to being able to route a given endpoint (in this case, the entry in the map)? 

If we think of the co-op close case, a new endpoint is only registered _after_ we send/recv shutdown. Once the channel has closed on-chain, there's no reason for the endpoints to hang around. 

I'd categorize this as the micro view: all unique execution instances have their own endpoint. We can contrast with the macro view wherein: all unique _messages_ have their own endpoint, with another msg router instance existing at that layer to further demultiplex. 

This PR implements micro routing. While I think if we want to have all message endpoints be registered (and permanent) at start up, we'd go with the macro route. At this point, not sure which is preferable. You can also build the macro version with the abstractions laid out in this PR (msg router hooks into another msg router).

Reply:
Yeah I'm definitely partial to going the macro route. It helps with exhaustiveness analysis. If we allow unregistering it means that there are messages that are only handleable at some times. To take your coop close case, I'm taking you to mean that ClosingSigned or ClosingComplete is only valid for the lifetime of the negotiation process. The approach you lay out here makes the structure of that logic implicit in the runtime execution of the code rather than explicit in the source where we have a check that says if our channel is in state X, message Y is not allowed.

Implicit logic is error prone for living codebases. It means that I have to keep certain contextual facts in my mind that are not reflected in the source. These assumptions may or may not change as time goes on. As a result, if you have not laid eyes on the code for a significant period of time, learning of this hidden assumption is not guaranteed. Contrast with the macro route where I can say "hmm let's see how message X from the bolt spec gets handled throughout the code" and then following it.

I do think this improves on the state of affairs but the dynamic nature of it is something I'm suspicious of.